### PR TITLE
Assignment 5

### PR DESCRIPTION
**Pull request by Fervidusletum**
Finished

### CODE REVIEW

**Comment by MeguminXIII** [comment]
I will Review

File: SecretSanta/test/SecretSanta.Web.Tests/Controllers/UsersControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            HttpClient client = Factory.CreateClient();
+
+            HttpResponseMessage response = await client.GetAsync("/users/");
+
+            response.EnsureSuccessStatusCode();
+            Assert.AreEqual(1, usersClient.GetAllAsyncInvokeCount);
+        }
+
+        [TestMethod]
+        public void Index_RetreivesValidDtoList_CreatesValidViewModelList()
```
Comment (MeguminXIII) [inline SecretSanta/test/SecretSanta.Web.Tests/Controllers/UsersControllerTests.cs:41]: Everything below this needs implementing

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        public void Put_GivenInvalidUser_Returns400()
+        {
+            Assert.Fail();
+        }
+
+        [TestMethod]
+        public void Put_GivenInvalidId_Returns404()
+        {
+            Assert.Fail();
+        }
```
Comment (MeguminXIII) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs:156]: needs implementing

**Review by MeguminXIII — COMMENTED** [review]
Assignment
Update the API project with the following
Add a new Dto namespace ✔
The UserController should only accept or return classes from the Dto namespace. Create any needed classes there and update the controller to use them. You will need to convert between the DTO classes and the relevant entities as needed. ✔
Add OpenAPI (swagger) endpoint ✔
Use the OpenAPI to generate a C# client with relevant interfaces ✔
Decorate all relevant API endpoints in the UserController with the appropriate ProducesResponseType ✔
Update the Web project with the following
Register the generated UserClients class using its interface with the Dependency Injection (DI) service. ✔
Update the UserController to take in the generated client as a dependency. ✔
Convert the UserController from using the MockData to using the client to get data from the API service. ✔
Update the controller to convert back and forth between the ViewModel classes and the API DTO classes as needed. ✔
Update the SecretSanta.Api.Tests with the following:
The controller tests should be interation tests not unit tests. ✔
All endpoints should be tested. ❌
some still not implemented, will put in a different comment

The UserController should use a test double for the IUserRepository to avoid needing to retest business and data logic. ✔
Create a new SecretSanta.Web.Tests project with the following:
It should be inside of the test directory. ✔
It should have nullable reference types enabled. ✔
It should have functional tests for all endpoints in the UserController that access the API. ❌
Most aren't implemented yet, will put in a separate comment

It should use a test double for the API service. This test double should be used to validate the controller's interaction with the API without actually calling the API. ✔
Extra Credit:

Rather than coping data between view models and DTOs, and DTOs and entities; use AutoMapper to handle it for you.❌

Looks good, just need to finish the tests

File: SecretSanta/src/SecretSanta.Api/Dto/UserDtoFnLn.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SecretSanta.Api.Dto
+{
+    public class UserDtoFnLn
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Api/Dto/UserDtoFnLn.cs:3]: nit: We typically avoid abbreviations unless they are common and very well known.

**Review by Keboo — COMMENTED** [review]
- Update the API project with the following
  - Add a new Dto namespace ✔
  - The `UserController` should only accept or return classes from the Dto namespace. Create any needed classes there and update the controller to use them. You will need to convert between the DTO classes and the relevant entities as needed. ✔
  - Add OpenAPI (swagger) endpoint ✔
  - Use the OpenAPI to generate a C# client with relevant interfaces ✔
  - Decorate all relevant API endpoints in the `UserController` with the appropriate [ProducesResponseType](https://docs.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-5.0#iactionresult-type) ✔
- Update the Web project with the following
  - Register the generated `UserClients` class using its interface with the Dependency Injection (DI) service. ✔
  - Update the `UserController` to take in the generated client as a dependency. ✔
  - Convert the `UserController` from using the `MockData` to using the client to get data from the API service. ✔
  - Update the controller to convert back and forth between the ViewModel classes and the API DTO classes as needed. ✔
- Update the SecretSanta.Api.Tests with the following:
  - The controller tests should be [integration tests](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-5.0) not unit tests. ✔
  - All endpoints should be tested.✔
  - The `UserController` should use a test double for the `IUserRepository` to avoid needing to retest business and data logic.✔
- Create a new SecretSanta.Web.Tests projecs with the following:
  - It should be inside of the test directory. ✔
  - It should have nullable reference types enabled. ✔
  - It should have functional tests for all endpoints in the `UserController` that access the API. ✔
  - It should use a test double for the API service. This test double should be used to validate the controller's interaction with the API without actually calling the API. ✔

**Pull request closed without merging**